modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java

```diff
@@ -140,31 +248,140 @@ public void twoCustomPropertiesTest(TestInfo testInfo) {
+     *
+     * @param testInfo Test information object.
+     */
+    @Test
+    public void testTwoNodesRestartDirect(TestInfo testInfo) {
+        twoNodesRestart(testInfo, true);
+    }
+
+    /**
+     * Starts two nodes and checks that the data are storing through restarts.
+     * Nodes restart in reverse order when they started at first.
+     *
+     * @param testInfo Test information object.
+     */
+    @Test
+    @Disabled("IGNITE-16034 Unblock a node start that happenes before Metastorage is ready")
+    public void testTwoNodesRestartReverse(TestInfo testInfo) {
+        twoNodesRestart(testInfo, false);
+    }
+
+    /**
+     * Starts two nodes and checks that the data are storing through restarts.
+     *
+     * @param testInfo Test information object.
+     * @param directOrder When the parameter is true, nodes restart in direct order, otherwise they restart in reverse order.
+     */
+    private void twoNodesRestart(TestInfo testInfo, boolean directOrder) {
+        Ignite ignite = startNode(testInfo, 0);
+
+        startNode(testInfo, 1);
+
+        createTableWithData(ignite, TABLE_NAME, 2);
+        createTableWithData(ignite, TABLE_NAME_2, 2);
+
+        stopNode(0);
+        stopNode(1);
+
+        if (directOrder) {
+            startNode(testInfo, 0);
+            ignite = startNode(testInfo, 1);
+        } else {
+            ignite = startNode(testInfo, 1);
+            startNode(testInfo, 0);
+        }
+
+        checkTableWithData(ignite, TABLE_NAME);
+        checkTableWithData(ignite, TABLE_NAME_2);
+    }
+
+    /**
+     * The test for node restart when there is a gap between the node local configuration and distributed configuration.
+     */
+    @Test
+    @Disabled("IGNITE-16718")
+    public void testCfgGap(TestInfo testInfo) {
+        final int nodes = 4;
+
+        for (int i = 0; i < nodes; i++) {
+            startNode(testInfo, i);
+        }
+
+        createTableWithData(CLUSTER_NODES.get(0), "t1", nodes);
+
+        String igniteName = CLUSTER_NODES.get(nodes - 1).name();
+
+        log.info("Stopping the node.");
+
+        IgnitionManager.stop(igniteName);
+
+        checkTableWithData(CLUSTER_NODES.get(0), "t1");
+
+        createTableWithData(CLUSTER_NODES.get(0), "t2", nodes);
+
+        log.info("Starting the node.");
+
+        Ignite newNode = IgnitionManager.start(igniteName, null, workDir.resolve(igniteName));
+
+        CLUSTER_NODES.set(nodes - 1, newNode);
+
+        checkTableWithData(CLUSTER_NODES.get(0), "t1");
+        checkTableWithData(CLUSTER_NODES.get(0), "t2");
+
+        checkTableWithData(CLUSTER_NODES.get(nodes - 1), "t1");
+        checkTableWithData(CLUSTER_NODES.get(nodes - 1), "t2");
+    }
+
+
+    /**
+     * Checks the table exists and validates all data in it.
+     *
+     * @param ignite Ignite.
+     * @param name Table name.
+     */
+    private void checkTableWithData(Ignite ignite, String name) {
```

Comment:
I believe that node recovery is a crucial logic of the app, so it worth to prepare explicit test plan for it. If you do not mind to do it, could you please share the test plan for node recovery within restart?